src/webui.c

```diff
@@ -77,6 +77,46 @@ fetch_wa_cookie(struct evhttp_request *req)
     return wa;
 }
 
+static void
+send_json_worker(struct evhttp_request *req, void *arg)
+{
+    struct evbuffer *buf = evhttp_request_get_output_buffer(req);
+    struct evkeyvalq *hdrs_out = NULL;
+    double mh[6] = {0};
+    char rig_id[MAX_RIG_ID] = {0};
+    char wnbuf[128] = {0}; /* buf was already in use */
+    char *wn = NULL;
+    const char *url = evhttp_request_get_uri(req);
+    url += 8; /* skip /worker/ */
+    strncpy(wnbuf, url, sizeof(wnbuf)-1);
+    wnbuf[strcspn(wnbuf, '/')] = 0;
+    wn = evhttp_uridecode(wnbuf, 1, NULL);
+    if (!wn)
+    {
+        /* send 404 response */
+        hdrs_out = evhttp_request_get_output_headers(req);
+        evbuffer_add_printf(buf, "{}");
+        evhttp_add_header(hdrs_out, "Content-Type", "application/json");
+        evhttp_send_reply(req, HTTP_NOTFOUND, "Not Found", buf);
+        return;
+    }
+    strncpy(rig_id, wn, MAX_RIG_ID-1);
+    free(wn);
+    const char *wa = fetch_wa_cookie(req);
+    if (wa)
+        worker_hr(mh, wa, rig_id);
```

Comment:
There's an odd logic here. If the user doesn't specify a worker name, you 404 (above, HTTP_NOTFOUND), but if they provide a worker name but no wallet address, you return 200 and a zero'ed object. Why not make sure both are provided for 200 and if either not provided 400 (HTTP_BADREQUEST)?

Ideally then you'd only 404 if _both_ arguments are provided but the worker really was not found. Thus have the function worker_hr return 0 (instead of void) as success (when found) else a non-zero for error (when not-found).

Reply:
```static void
send_json_worker(struct evhttp_request *req, void *arg)
{
    struct evbuffer *buf = evhttp_request_get_output_buffer(req);
    struct evkeyvalq *hdrs_out = NULL;
    double mh[6] = {0};
    char rig_id[MAX_RIG_ID] = {0};
    char wnbuf[128] = {0};
    char *wn = NULL;
    const char *url = evhttp_request_get_uri(req);
    url += 8; 
    strncpy(wnbuf, url, sizeof(wnbuf)-1);
    wnbuf[strcspn(wnbuf, '/')] = 0;
    wn = evhttp_uridecode(wnbuf, 1, NULL);
    const char *wa = fetch_wa_cookie(req);
    if (!wn || !wa)
    {
        hdrs_out = evhttp_request_get_output_headers(req);
        evbuffer_add_printf(buf, "{}");
        evhttp_add_header(hdrs_out, "Content-Type", "application/json");
        evhttp_send_reply(req, HTTP_NOTFOUND, "Not Found", buf);
        return;
    }
    strncpy(rig_id, wn, MAX_RIG_ID-1);
    free(wn);
    worker_hr(mh, wa, rig_id);
    evbuffer_add_printf(buf, "{"
            "\"miner_hashrate_stats\":["
            "%"PRIu64",%"PRIu64",%"PRIu64","
            "%"PRIu64",%"PRIu64",%"PRIu64"]"
            "}",
            (uint64_t)mh[0], (uint64_t)mh[1], (uint64_t)mh[2],
            (uint64_t)mh[3], (uint64_t)mh[4], (uint64_t)mh[5]);
    hdrs_out = evhttp_request_get_output_headers(req);
    evhttp_add_header(hdrs_out, "Content-Type", "application/json");
    evhttp_send_reply(req, HTTP_OK, "OK", buf);
}```

Reply:
bad request implies the user has provided invalid arguments.
not found implies the object could exist but not located.

I'm not sure what is the best type to send back I will defer to you on updating it to bad request or leaving with not found.

Reply:
Both wallet address and worker name are required parameters. And If a required parameter is not supplied, the response should be BADREQUEST. If both parameters _are_ supplied but there is no account-worker combination found, the response should be NOTFOUND. Hence my comment that worker_hr should return a status/error to indicate whether the account-worker was found.

Reply:
been busy will have time to update per your requests in the next couple of weeks.

Reply:
Thanks